pkg/storegateway/bucket.go

```diff
@@ -1241,34 +1237,101 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 // so we could also intersect those with each label's postings being each one non empty and leading to the same result.
-func blockLabelValues(ctx context.Context, indexr *bucketIndexReader, labelName string, matchers []*labels.Matcher, logger log.Logger, stats *safeQueryStats) ([]string, error) {
+func blockLabelValues(ctx context.Context, b *bucketBlock, postingsStrategy postingsSelectionStrategy, maxSeriesPerBatch int, labelName string, matchers []*labels.Matcher, logger log.Logger, stats *safeQueryStats) ([]string, error) {
+	indexr := b.indexReader(postingsStrategy) // this strategy doesn't make a difference, since we won't be using ExpandedPostings with this reader
+	defer runutil.CloseWithLogOnErr(b.logger, indexr, "close block index reader")
+
 	values, ok := fetchCachedLabelValues(ctx, indexr.block.indexCache, indexr.block.userID, indexr.block.meta.ULID, labelName, matchers, logger)
 	if ok {
 		return values, nil
 	}
 
 	// TODO: if matchers contains labelName, we could use it to filter out label values here.
-	allValues, err := indexr.block.indexHeaderReader.LabelValues(labelName, "", nil)
+	allValuesPostingOffsets, err := indexr.block.indexHeaderReader.LabelValuesOffsets(labelName, "", nil)
 	if err != nil {
 		return nil, errors.Wrap(err, "index header label values")
 	}
 
 	if len(matchers) == 0 {
-		storeCachedLabelValues(ctx, indexr.block.indexCache, indexr.block.userID, indexr.block.meta.ULID, labelName, matchers, allValues, logger)
-		return allValues, nil
+		values = extractLabelValues(allValuesPostingOffsets)
+		storeCachedLabelValues(ctx, indexr.block.indexCache, indexr.block.userID, indexr.block.meta.ULID, labelName, matchers, values, logger)
+		return values, nil
 	}
+	strategy := &labelValuesPostingsStrategy{
+		delegate:     postingsStrategy,
+		postingLists: allValuesPostingOffsets,
+	}
+	indexr = b.indexReader(strategy)
+	defer runutil.CloseWithLogOnErr(b.logger, indexr, "close block index reader")
 
-	p, pendingMatchers, err := indexr.ExpandedPostings(ctx, matchers, stats)
+	postings, pendingMatchers, err := indexr.ExpandedPostings(ctx, matchers, stats)
 	if err != nil {
 		return nil, errors.Wrap(err, "expanded postings")
 	}
+	if len(pendingMatchers) > 0 || strategy.preferSeriesToPostings(postings) {
+		values, err = labelValuesFromSeries(ctx, labelName, maxSeriesPerBatch, pendingMatchers, indexr, b, postings, stats)
+	} else {
+		values, err = labelValuesFromPostings(ctx, labelName, indexr, allValuesPostingOffsets, postings, stats)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	storeCachedLabelValues(ctx, indexr.block.indexCache, indexr.block.userID, indexr.block.meta.ULID, labelName, matchers, values, logger)
+	return values, nil
+}
+
+func labelValuesFromSeries(ctx context.Context, labelName string, seriesPerBatch int, pendingMatchers []*labels.Matcher, indexr *bucketIndexReader, b *bucketBlock, postings []storage.SeriesRef, stats *safeQueryStats) ([]string, error) {
```

Comment:
[nit] I would clarify this too.

```suggestion
func labelValuesFromSeries(ctx context.Context, labelName string, seriesPerBatch int, pendingMatchers []*labels.Matcher, indexr *bucketIndexReader, b *bucketBlock, matchersPostings []storage.SeriesRef, stats *safeQueryStats) ([]string, error) {
```